### DISCUSSION TITLE
[ramda] Un-curried `pick` is too weakly typed

### DISCUSSION THREAD

**Discussion by johannesleander** [DefinitelyTyped/DefinitelyTyped#63571] [Issues with a @types package]
While using [Ramda](https://ramdajs.com/)'s un-curried `pick` function, I noticed it doesn't render typing errors when one is attempting to access keys not present in a given object.

[Here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/0d302353cd161804d1904991e6e617d3066962dd/types/ramda/index.d.ts#L3929) are the current types for @types/ramda's `pick` function.

```typescript
const charlie = R.pick([ 'charlie' ], { 'annie': true, 'bob': true }) // ✅ Doesn't result in a TS error
console.log(charlie) // undefined
```

Instead, I was expecting the function to behave like this:
```typescript
const charlie = R.pick([ 'charlie' ], { 'annie': true, 'bob': true }) // ❌
//                          ^^^^^  Type 'charlie' is not assignable to type 'annie' | 'bob'
console.log(charlie) // undefined
```

In a project of mine I've created a proxy type that more strictly limits which keys can be provided.
```typescript
type StrictPick = <T, K extends keyof T>(
  keys: readonly K[],
  value: T
) => Pick<T, Exclude<keyof T, Exclude<keyof T, K>>>
const strictPick: StrictPick = R.pick

strictPick([ 'charlie' ], { 'annie': true, 'bob': true }) // ❌
```

Now - I do see how this could be a problem from a functional purity stand-point. When attempting to curry this function using `Ramda.curry` the generic types are unable to resolve and will render unwanted TS errors.
```typescript
const curriedStrictPick = R.curry(strictPick)

curriedStrictPick([ 'charlie' ])({ 'annie': true, 'bob': true, 'charlie': true }) // ❌
//                    ^^^^^  Type 'string' is not assignable to type 'never'
```

I currently don't have a solution to the currying problem and would understand why such a limitation could hold this proposal back. 

Any thoughts on how to allow for stricter typing of Ramda's pick?

[a comment by an automated account is not shown]

**Comment by couzic** [comment]
Hello @johannesleander,

The code:
```typescript
function pick<T, K extends string | number | symbol>(
  names: readonly K[],
  obj: T
): { [P in keyof T as P extends K ? P : never]: T[P] } {
  return R.pick(names, obj) as any;
}

const oldCharlie = R.pick(["charlie"], { annie: true, bob: true }); // Pick<{ annie: boolean, bob: boolean }, never>
console.log(oldCharlie); // {}
const newCharlie = pick(["charlie"], { annie: true, bob: true }); // {}

const oldAnnie = R.pick(["annie"], { annie: true, bob: true }); // Pick<{ annie: boolean, bob: boolean }, "annie">
console.log(oldAnnie); // { annie: true }
const newAnnie = pick(["annie"], { annie: true, bob: true }); // { annie: boolean }
```
Same solution for curried version:
```typescript
function pick<K extends string | number | symbol>(
  names: readonly K[]
): <T>(obj: T) => { [P in keyof T as P extends K ? P : never]: T[P] } {
  return R.pick(names) as any;
}

const oldCharlie = R.pick(["charlie"])({ annie: true, bob: true }); // Pick<{ annie: boolean, bob: boolean }, never>
console.log(oldCharlie); // {}
const newCharlie = pick(["charlie"])({ annie: true, bob: true }); // {}

const oldAnnie = R.pick(["annie"])({ annie: true, bob: true }); // Pick<{ annie: boolean, bob: boolean }, "annie">
console.log(oldAnnie); // { annie: true }
const newAnnie = pick(["annie"])({ annie: true, bob: true }); // { annie: boolean }
```

**Comment by couzic** [comment]
@johannesleander to give you a more complete answer, the change you are proposing would be a contract change that would not fit the runtime code, which is permissive enough to allow keys that are not present in the object.
However your question gave me the idea of the code improvement I shared, I just created a PR so it becames available for everyone:
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/63809

**Reply by johannesleander** [reply]
Thanks for clarifying. That makes sense, despite the drawback of a poorer development experience due to a lack of e.g. autocompletion. The types reflecting the runtime contract is more important than autocompletion.
